### PR TITLE
Move csync to src/ and compile most of it at C++

### PR DESCRIPTION
**Pull request by jturcotte**
Most changes are mechanical and don't change any logic in the code.

This will make it easier to hack and refactor csync by allowing us to use QByteArray and smart pointers to manage memory allocations. Eventually this would also allow us to use QString and QFile in csync rather than having to convert to UTF-8 in and out of the library boundary.

### CODE REVIEW

**Comment by jturcotte** [comment]
It would be nice if we could manage to get this in master before we branch for 2.4.

File: src/csync/csync_update.cpp [hunk 1]
```diff
@@ -648,13 +651,13 @@ int csync_ftw(CSYNC *ctx, const char *uri, csync_walker_fn fn,
...
   int res = 0;
 
+  bool do_read_from_db = (ctx->current == REMOTE_REPLICA && ctx->remote.read_from_db);
+
   if (!depth) {
     mark_current_item_ignored(ctx, previous_fs, CSYNC_STATUS_INDIVIDUAL_TOO_DEEP);
     goto done;
   }
 
-  bool do_read_from_db = (ctx->current == REMOTE_REPLICA && ctx->remote.read_from_db);
```
Comment (guruz) [inline src/csync/csync_update.cpp]: Just commenting to make you aware of it..
I guess this reordering has no effect..

File: src/csync/csync_exclude.cpp [hunk 2]
```diff
@@ -331,7 +331,6 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(c_strlist_t *excludes, const ch
         goto out;
     }
 
-    c_strlist_t *path_components = NULL;
```
Comment (guruz) [inline src/csync/csync_exclude.cpp]: No real reason to move this up? But not an issue, just leave it..

File: src/libsync/utility.cpp [hunk 3]
```diff
@@ -275,7 +275,7 @@ void Utility::usleep(int usec)
 
 bool Utility::fsCasePreserving()
 {
-#ifndef WITH_TESTING
+#ifdef WITH_TESTING
```
Comment (guruz) [inline src/libsync/utility.cpp]: heh

File: config.h.in [hunk 4]
```diff
@@ -24,6 +24,6 @@
 #cmakedefine SYSCONFDIR "@SYSCONFDIR@"
 #cmakedefine SHAREDIR "@SHAREDIR@"
 
-#cmakedefine WITH_UNIT_TESTING 1
+#cmakedefine WITH_TESTING 1
```
Comment (guruz) [inline config.h.in]: Make sure the Jenkinsfile would still run the tests.
Also make sure make test would still run the tests

File: src/csync/csync_update.cpp (hunk 1 above)
Comment (jturcotte) [reply src/csync/csync_update.cpp]: C++ complains if a goto crosses a variable declaration.

File: src/libsync/utility.cpp (hunk 3 above)
Comment (jturcotte) [reply src/libsync/utility.cpp]: I did it wrong initially and it worked since the right config.h wasn't included here.

File: config.h.in (hunk 4 above)
Comment (jturcotte) [reply config.h.in]: 👍

Comment (guruz) [reply config.h.in]: Now I'm wondering why we have UNIT_TESTING and WITH_TESTING

Comment (jturcotte) [reply config.h.in]: It could be renamed, UNIT_TESTING is the cmake variable.

**Pull request merged**